automation/run/store.go

```diff
@@ -0,0 +1,274 @@
+// events and updating the associated the run.State accordingly
+type Store interface {
+	// Put adds a run state to the Store. If the given RunID is empty, it
+	// creates a new run.State. If RunID is not empty, it updates the
+	// associated run.
+	Put(r *State) (*State, error)
+	// Get gets the associated run.State
+	Get(id string) (*State, error)
+	// Count returns the number of runs for a given workflow.ID
+	Count(wid workflow.ID) (int, error)
+	// List lists all the runs associated with the workflow.ID in reverse
+	// chronological order
+	List(wid workflow.ID, lp params.List) ([]*State, error)
+	// GetLatest returns the most recent run associated with the workflow id
+	GetLatest(wid workflow.ID) (*State, error)
+	// GetStatus returns the status of the latest run based on the
+	// workflow.ID
+	GetStatus(wid workflow.ID) (Status, error)
+	// ListByStatus returns a list of run.State entries with a given status
+	// looking only at the most recent run of each Workflow
+	ListByStatus(s Status, lp params.List) ([]*State, error)
+}
+
+// EventAdder is an extension interface that optimizes writing an event to a run
+// Result should equal to calling:
+//   run := store.Get(id)
+//   run.AddTransformEvent(e)
+//   store.Put(run)
+type EventAdder interface {
+	Store
+	// AddEvent writes an event to the store, attaching it to an existing stored
+	// run state
+	AddEvent(id string, e event.Event) error
+}
+
+// MemStore is an in memory representation of a Store
+type MemStore struct {
+	mu              sync.Mutex
+	bus             event.Bus
+	workflows       map[workflow.ID]*workflowMeta
+	runs            map[string]*State
+	handleRunEvents event.Handler
+}
+
+type workflowMeta struct {
+	count  int
+	runIDs []string
+}
+
+func newWorkflowMeta() *workflowMeta {
+	return &workflowMeta{
+		count:  0,
+		runIDs: []string{},
+	}
+}
+
+var (
+	_ Store      = (*MemStore)(nil)
+	_ EventAdder = (*MemStore)(nil)
+)
+
+// NewMemStore returns a MemStore
+func NewMemStore() *MemStore {
+	s := &MemStore{
+		workflows: map[workflow.ID]*workflowMeta{},
+		runs:      map[string]*State{},
+	}
+	return s
+}
+
+// Put adds a run.State to a MemStore
+func (s *MemStore) Put(r *State) (*State, error) {
+	if r == nil {
+		return nil, fmt.Errorf("run is nil")
+	}
+	run := r.Copy()
+	if run.ID != "" {
+		if fetchedR, err := s.Get(run.ID); !errors.Is(err, ErrNotFound) {
+			if fetchedR.WorkflowID != run.WorkflowID {
+				return nil, ErrPutWorkflowIDMismatch
+			}
+		}
+	} else {
+		run.ID = NewID()
+	}
+	if err := run.Validate(); err != nil {
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.workflows[run.WorkflowID]
+	if !ok {
+		wf := newWorkflowMeta()
+		s.workflows[run.WorkflowID] = wf
+	}
+	_, ok := s.runs[run.ID]
+	if !ok {
+		s.workflows[run.WorkflowID].count++
+		runIDs := s.workflows[run.WorkflowID].runIDs
+		s.workflows[run.WorkflowID].runIDs = append(runIDs, run.ID)
+	}
+	s.runs[run.ID] = run
+	return run, nil
```

Comment:
@b5 Okay, so started implementing pulling out adjusting the run count and workflow in a separate method and adjusting the store Interface... it felt a bit like overkill & kind of awkward.

Basically, felt like the orchestrator was having opinions about the way that `run.Store` should be implemented.

EG, in a future SQL implementation, having a second call to something like `UpdateCount` or `UpdateOtherThings` would make no sense. If we really want to control that something different happens the first time you initiate a run, I'd rather separate the concerns by having `runStore.Create` method, and use `Put` for updates, rather than a second call after `Put` for special behavior. 

I think the only reason we had this bug was because I made a mistake while coding, not because there is something inherently wrong in how we've designed it.

Thoughts?

Reply:
yeah, looking at this implementation, and how we're working with the store up in the orchestrator, I think a `Create` method makes sense. I feel like we're relying too heavily on an empty `ID` field to determine user intent, where a `Create` method would make it explicit that you expect to, well, create a thing.

So, I'd vote for the former: let's add `run.Store.Create`. Let's reposition existing spec tests that work the empty-id-Put call to test the create method. I think it'll clean a bunch of stuff up